File: cellab_devkit-1.0 (1)/ruledefs/brailife.js
    /*

            This rule runs Life and Brain in parallel and lets them
            interact only within a certain masked region.  In this region,
            firing Brain cells turn on Life cells, and firing Life cells
            keep Brain cells from turning on.
	    
	    We use the eight bits of state as follows:
		Bit #0 is used to show either the Brain
        	       or the Life bit to neighbors;
		Bit #1 is the Life bit,
		Bit #2 is the firing Brain bit,
		Bit #3 is the refractory Brain bit,
		Bit #4 is the mask bit,and
		Bit #7 is the cycle bit.

    */

    rule.worldtype = 1;
    /* The starting BraiLife pattern has all bit 7s set to 0
       (for synchronized cycles), has two adjacent cells of
        plane #2 turned on to start Brain, and has a big disk
        mask in plane #4. */
    rule.patreq = "brailife";
    /* The brailife.jcc colorpalette looks at bits 4,3,2, &
       1.  I got my colorpalette by disabling planes 0,5,6,7 of
       the Default.JCC and saving it as brailife.jcc. */
    rule.palreq = "brailife";
    
    function brailife(oldstate,     nw, n  , ne,
                                    w, self, e,
                                    sw, s  , se) {
        var r = 0, l, newL, b, newB, bR, newBR,
            mask, cycle, newCycle, eightSum;

        eightSum = nw + n + ne + e + se + s + sw + w;
        cycle = (oldstate >> 7) & 1;
        mask = (oldstate >> 4) & 1;
        bR = (oldstate >> 3) & 1;
        b = (oldstate >> 2) & 1;
        l = (oldstate >> 1) & 1;

        if (cycle == 0) {

            //  Life update cycle

            if ((eightSum == 3) || ((eightSum == 2) && (l == 1))) {
                newL = 1;
            } else {
                newL = 0;
            }
            //  Turned on by firing Brain cells within region of mask
            if (mask == 1 && b == 1) {
                newL = 1;
            }
            newCycle = 1;
            r = (newCycle << 7) | (mask << 4) | (bR << 3) |
                (b << 2) | (newL << 1) | b;
        } else {

            //  Brain update cycle

            if (((bR == 0) && (b == 0)) && (eightSum == 2)) {
                newB = 1;
            } else {
                newB = 0;
            }
            //  Turned off by firing Life cells within region of mask
            if (l == 1 && mask == 1) {
                newB = 0;
            }
            newBR = b;
            newCycle = 0;
            r = (newCycle << 7) | (mask << 4) | (newBR << 3) |
                (newB << 2) | (l << 1) | l;
        }
        return r;
    }

File: references/cellab_devkit-1.0 (1)/rules_java/brailife.java
/*

        This rule runs Life and Brain in parallel and lets them
        interact only within a certain masked region.  In this region,
        firing Brain cells turn on Life cells, and firing Life cells
        keep Brain cells from turning on.

*/

class Brailife extends ruletable {

    void jcruleModes() {
        /*  The BraiLife.CAC palette looks at bits 4,3,2, & 1.
            Practically every combo has a meaning, so suppose I get my
            palette simply by disabling planes 0,5,6,7.  */
        setPaletteRequest("BraiLife");
        /*  The starting BraiLife pattern has all bit 7s set to 0 (for
            synchronized cycles), has two adjacent cells on in plane
            #2 turned on to start Brain, and has a big disk mask in
            plane 4.  */
        setPatternRequest("BraiLife");
    }

    int jcrule(int oldstate) {
        int r = 0, L, NewL, B, NewB, BR, NewBR,
            Mask, Cycle, NewCycle, EightSum;

        EightSum = nw + n + ne + e + se + s + sw + w;
        Cycle = (oldstate >> 7) & 1;
        Mask = (oldstate >> 4) & 1;
        BR = (oldstate >> 3) & 1;
        B = (oldstate >> 2) & 1;
        L = (oldstate >> 1) & 1;

        if (Cycle == 0) {

            //  Life update cycle

            if ((EightSum == 3) || ((EightSum == 2) && (L == 1))) {
                NewL = 1;
            } else {
                NewL = 0;
            }
            //  Turned on by firing Brain cells within region of mask
            if (Mask == 1 && B == 1) {
                NewL = 1;
            }
            NewCycle = 1;
            r = (NewCycle << 7) | (Mask << 4) | (BR << 3) |
                (B << 2) | (NewL << 1) | B;
        } else {

            //  Brain update cycle

            if (((BR == 0) && (B == 0)) && (EightSum == 2)) {
                NewB = 1;
            } else {
                NewB = 0;
            }
            //  Turned off by firing Life cells within region of mask
            if (L == 1 && Mask == 1) {
                NewB = 0;
            }
            NewBR = B;
            NewCycle = 0;
            r = (NewCycle << 7) | (Mask << 4) | (NewBR << 3) |
                (NewB << 2) | (L << 1) | L;
        }
        return r;
    }
}

public class brailife {
    public static void main(String args[]) {
        (new Brailife()).generateRuleFile("brailife");
    }
}
